### ISSUE TITLE
The ability to encode using Huffman encoding should be exposed

### ISSUE THREAD

**Issue by mtrudel** [elixir-mint/hpax#3]
The functionality is present (though I'm unsure how tested it is), but the relevant flags are always set to false with no way for a caller to change them (see, for example: https://github.com/elixir-mint/hpax/pull/1/files#diff-4f483d3ec827149f45394513652cebbf629a48f1023bd62153f775ebd27a365eR224)

**Comment by whatyouhide** [comment]
@mtrudel do you mean exposing `HPAX.Huffman` itself, or expose the ability to use the Huffman encoding in HPAX?

**Comment by mtrudel** [comment]
It's been a while (and I don't have my paper notes from that long ago with me at the moment), but I believe the desire here was to be able to provide a knob to configure which of the possible encodings that HPAX uses & how it uses them (IIRC, RFC7541 allows a number of encodings up to and including simple passthrough). I'll load all this back into my brain once I'm home next week and add some more coherent thoughts here.

**Comment by mtrudel** [comment]
> Before doing this I would like to see some experiments/benchmarking to see how useful it would be. Either way it should be an option if we support it. What do other HTTP/2 libraries do here?

Agreed. I can pretty easily see what clients do in this situation just by instrumenting the decoding logic in Bandit's use of HPax. I'll do a bit of digging on the following and report back:

* What various clients do
* What Cowboy does
* How much of a difference encoding would make in some common exchanges

**Comment by mtrudel** [comment]
tl;dr: Huffman encoding seems to make a pretty big difference in practice. I'm going to hold off on recommendations just now, as there's more to the story in my next few comments.

### What Clients Do

I ran a simple single 'hello world' GET request and found the following (details follow):

* 91% of all string literals were sent Huffman encoded
* Of the 5 identity encoded literals, none would have been smaller had they been Huffman encoded
* Safari saves 61 bytes per request (197 bytes vs 258 bytes), a ratio of 0.76
* Chrome saves 155 bytes per request (491 bytes vs 646 bytes), a ratio of 0.76
* Firefox saves 107 bytes per request (317 bytes vs 424 bytes), a ratio of 0.75
* Curl saves 7 bytes per request (18 bytes vs 25 bytes), a ratio of 0.72

Note that as these numbers are for *very* simple requests, with simple paths & hostnames, minimal application-level headers, etc. I strongly suspect in practice that these numbers would be even more compelling.

### Cowboy

A quick dive into the Cowboy source shows that they always encode using Huffman encoding, and do not provide a user-level override to configure it.

### Data

<details>
  <summary>Request Data</summary>

User Agent | algorithm | Compression | Encoded bytes | Decoded bytes | Header |   |   |  
-- | -- | -- | -- | -- | -- | -- | -- | --
Safari | huffman | 71% | 10 | 14 | localhost:4040 |   |   |  
Safari | huffman | 81% | 17 | 21 | lsid=XcBoGMLmuvKhhtMZ |   |   |  
Safari | huffman | 76% | 48 | 63 | text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8 |   |   |  
Safari | huffman | 75% | 88 | 117 | Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.4 Safari/605.1.15 |   |   |  
Safari | huffman | 81% | 21 | 26 | en-CA,en-US;q=0.9,en;q=0.8 |   |   |  
Safari | huffman | 76% | 13 | 17 | gzip, deflate, br |   |   |  
Chrome | huffman | 71% | 10 | 14 | localhost:4040 |   |   |  
Chrome | huffman | 78% | 7 | 9 | max-age=0 |   |   |  
Chrome | huffman | 78% | 7 | 9 | sec-ch-ua |   |   |  
Chrome | huffman | 86% | 56 | 65 | Chromium";v="112" | "Google Chrome";v="112" | "Not:A-Brand";v="99" |  
Chrome | huffman | 69% | 11 | 16 | sec-ch-ua-mobile |   |   |  
Chrome | literal | 100% | 2 | 2 | ?0 |   |   |  
Chrome | huffman | 72% | 13 | 18 | sec-ch-ua-platform |   |   |  
Chrome | literal | 100% | 7 | 7 | macOS" |   |   |  
Chrome | huffman | 72% | 18 | 25 | upgrade-insecure-requests |   |   |  
Chrome | literal | 100% | 1 | 1 | 1 |   |   |  
Chrome | huffman | 76% | 89 | 117 | Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36 |   |   |  
Chrome | huffman | 75% | 101 | 135 | text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7 |   |   |  
Chrome | huffman | 71% | 10 | 14 | sec-fetch-site |   |   |  
Chrome | huffman | 75% | 3 | 4 | none |   |   |  
Chrome | huffman | 71% | 10 | 14 | sec-fetch-mode |   |   |  
Chrome | huffman | 75% | 6 | 8 | navigate |   |   |  
Chrome | huffman | 71% | 10 | 14 | sec-fetch-user |   |   |  
Chrome | literal | 100% | 2 | 2 | ?1 |   |   |  
Chrome | huffman | 71% | 10 | 14 | sec-fetch-dest |   |   |  
Chrome | huffman | 75% | 6 | 8 | document |   |   |  
Chrome | huffman | 76% | 13 | 17 | gzip, deflate, br |   |   |  
Chrome | huffman | 83% | 19 | 23 | en-US,en;q=0.9,fr;q=0.8 |   |   |  
Chrome | huffman | 75% | 9 | 12 | /favicon.ico |   |   |  
Chrome | huffman | 77% | 49 | 64 | image/avif,image/webp,image/apng,image/svg+xml,image/*,*/*;q=0.8 |   |   |  
Chrome | huffman | 73% | 8 | 11 | same-origin |   |   |  
Chrome | huffman | 71% | 5 | 7 | no-cors |   |   |  
Chrome | huffman | 80% | 4 | 5 | image |   |   |  
Chrome | huffman | 74% | 17 | 23 | https://localhost:4040/ |   |   |  
Firefox | huffman | 100% | 1 | 1 | / |   |   |  
Firefox | huffman | 71% | 10 | 14 | localhost:4040 |   |   |  
Firefox | huffman | 74% | 62 | 84 | Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:109.0) Gecko/20100101 Firefox/112.0 |   |   |  
Firefox | huffman | 75% | 64 | 85 | text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8 |   |   |  
Firefox | huffman | 81% | 21 | 26 | en-CA,en-US;q=0.7,en;q=0.3 |   |   |  
Firefox | huffman | 76% | 13 | 17 | gzip, deflate, br |   |   |  
Firefox | huffman | 72% | 18 | 25 | upgrade-insecure-requests |   |   |  
Firefox | huffman | 100% | 1 | 1 | 1 |   |   |  
Firefox | huffman | 71% | 10 | 14 | sec-fetch-dest |   |   |  
Firefox | huffman | 75% | 6 | 8 | document |   |   |  
Firefox | huffman | 71% | 10 | 14 | sec-fetch-mode |   |   |  
Firefox | huffman | 75% | 6 | 8 | navigate |   |   |  
Firefox | huffman | 71% | 10 | 14 | sec-fetch-site |   |   |  
Firefox | huffman | 75% | 3 | 4 | none |   |   |  
Firefox | huffman | 71% | 10 | 14 | sec-fetch-user |   |   |  
Firefox | huffman | 100% | 2 | 2 | ?1 |   |   |  
Firefox | huffman | 100% | 2 | 2 | te |   |   |  
Firefox | huffman | 75% | 6 | 8 | trailers |   |   |  
Firefox | huffman | 75% | 9 | 12 | /favicon.ico |   |   |  
Firefox | huffman | 76% | 19 | 25 | image/avif,image/webp,*/* |   |   |  
Firefox | huffman | 74% | 17 | 23 | https://localhost:4040/ |   |   |  
Firefox | huffman | 80% | 4 | 5 | image |   |   |  
Firefox | huffman | 71% | 5 | 7 | no-cors |   |   |  
Firefox | huffman | 73% | 8 | 11 | same-origin |   |   |  
Curl | huffman | 71% | 10 | 14 | localhost:4040 |   |   |  
Curl | huffman | 73% | 8 | 11 | curl/7.79.1 |   |   |  
Curl | literal | 100% | 3 | 3 | */* |   |   |  

</details>

<details>
  <summary>Test Setup</summary>

```
defmodule MyPlug do
  import Plug.Conn

  def init(options) do
    options
  end

  def call(conn, _opts) do
    conn
    |> put_resp_content_type("text/plain")
    |> send_resp(200, "Hello world")
  end
end

Bandit.start_link(plug: MyPlug, scheme: :https, keyfile: "/Users/mat/Code/bandit/test/support/key.pem", certfile: "/Users/mat/Code/bandit/test/support/cert.pem")
```

</details>

**Comment by whatyouhide** [comment]
@mtrudel awesome, this is a fantastic dive, thank you so much! One thing I ='d love to know is this: how does this affect performance on our end? It seems like payload size is definitely better, but I'd like to know at what performance cost.

**Comment by mtrudel** [comment]
Huffman is (unsurprisingly) quite a bit slower; about 4x or so. In absolute terms though, it works out to about an even win:

The extra time to do Huffman encoding works out to about 5us per encoding run. The test data set encodes to 212 bytes Huffman encoded vs 273 bytes literally encoded (0.77x). Assuming a 100 megabit network, those extra 51 bytes will take about 5us to transmit. So from a time perspective (and with some WILDLY HANDWAVING assumptions :) we net out about the same in both cases, just trading CPU time off for network time. Of course these numbers would vary based on real-world header sizes, varying network & IO speeds, server vs my laptop, etc. 

<details>
  <summary>Results</summary>

```
Operating System: macOS
CPU Information: Apple M1
Number of Available Cores: 8
Available memory: 16 GB
Elixir 1.14.3
Erlang 25.1

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 10 s
memory time: 2 s
reduction time: 0 ns
parallel: 1
inputs: none specified
Estimated total run time: 28 s

Benchmarking huffman ...
Benchmarking literal ...

Name              ips        average  deviation         median         99th %
literal      702.19 K        1.42 μs  ±1593.15%        1.29 μs        1.54 μs
huffman      161.56 K        6.19 μs   ±212.25%        5.63 μs       10.75 μs

Comparison:
literal      702.19 K
huffman      161.56 K - 4.35x slower +4.77 μs

Memory usage statistics:

Name       Memory usage
literal         3.92 KB
huffman        14.75 KB - 3.76x memory usage +10.83 KB
```

</details>

<details>
  <summary>Test Setup</summary>

```
Mix.install([{:hpax, path: "."}, :benchee])

headers = [
  {:store, ":method", "GET"},
  {:store, ":scheme", "https"},
  {:store, ":path", "/"},
  {:store, ":authority", "localhost:4040"},
  {:store, "cookie", "lsid=XcBoGMLmuvKhhtMZ"},
  {:store, "accept",
   "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8"},
  {:store, "user-agent",
   "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.4 Safari/605.1.15"},
  {:store, "accept-language", "en-CA,en-US;q=0.9,en;q=0.8"},
  {:store, "accept-encoding", "gzip, deflate, br"}
]

literal_tbl = %{HPAX.Table.new(1024) | huffman: :never}
huffman_tbl = HPAX.Table.new(1024)

Benchee.run(
  %{
    "literal" => fn -> HPAX.encode(headers, literal_tbl) end,
    "huffman" => fn -> HPAX.encode(headers, huffman_tbl) end
  },
  time: 10,
  memory_time: 2
)
```

</details>

**Comment by mtrudel** [comment]
(The above work is based on the WiP at https://github.com/mtrudel/hpax)

**Comment by mtrudel** [comment]
Yep, I'll get that WiP tidied up!
